### PR TITLE
chore(vpat): add VPAT report generation POC

### PR DESCRIPTION
**Pull request by dtopalov**
# PR Summary: VPAT Generation & Accessibility Infrastructure

## VPAT Report Generator

Added a VPAT® 2.5Rev WCAG Edition report generator to the `@progress/kendo-themes-html` package.

**New files in vpat:**
- vpat-generator.ts — generates Markdown reports from test data and configuration
- `default-config.ts` — baseline VPAT config with known exceptions, general compliance statements, and WCAG criteria mappings
- `wcag-criteria.ts` — WCAG 2.0/2.1/2.2 criteria reference (Levels A, AA, AAA)
- `types.ts` — TypeScript interfaces for config, overrides, reports, and test analysis
- index.ts — public exports (`generateVpatReport`, types)

**CLI wrapper (vpat):**
- generate-vpat.mjs — CLI for generating Markdown reports with `--overrides`, `--a11y-report`, `--contrast-report`, `--out` arguments
- README.md — full usage documentation (programmatic API, CLI, PDF generation, overrides format)

**Unit tests:**
- vpat-generator.test.ts — 39 tests covering report generation, config merging, override behavior, WCAG criteria mapping, and edge cases

## A11y Test Infrastructure

**Contrast test script (test-contrast.mjs):**
- New script for running contrast ratio tests with JSON report output

**A11y test updates (test-a11y-unified.mjs):**
- Updated to support JSON report output alongside existing formats

**CI workflow updates:**
- `_test-a11y-specs.yml` — runs headless via `HEADLESS=true`
- `_test-contrast.yml` — new reusable contrast test workflow

## Component Accessibility Fixes

- **Signature** — added full accessibility support: `aria-label` on action buttons, `role="img"` + `aria-label` + `tabIndex` on canvas, full `ariaSpec` with 4 rules
- **SplitButton** — fixed `aria-disabled` being overwritten by child Button's explicit attribute (now uses Button's `ariaDisabled` prop)
- **MenuButton** — same `aria-disabled` fix as SplitButton
- **SmartBox** — fixed `aria-pressed` being overwritten (now uses Button's `togglable` and `selected` props)
- **PanelBar** — fixed `aria-hidden` being overwritten on group element (conditionally applies only when explicitly provided)
- **TileLayout** — removed deprecated WAI-ARIA 1.2+ attributes (`aria-grabbed`, `aria-dropeffect`) from ariaSpec, rendering, and test HTML files

## Cleanup

- Removed `aria/` folder — 116 markdown spec files deleted. The `ariaSpec` object on each component's `.spec.tsx` is now the single source of truth
- Updated all references in copilot-instructions.md, `accessibility.prompt.md`, `html.instructions.md`
- Updated .gitignore with test output patterns

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (epetrow) [inline .github/workflows/ci.yml:1]: It seems like by default the GH Compare API returns up to 300 files so if there are more than that there might be some missed changes. Also if the API call fails wouldn't `FILES and `MATCH` become empty causing the a11y tests to get silently skipped?

File: .github/workflows/nightly-a11y.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            echo "changed=false" >> $GITHUB_OUTPUT
+            echo "No report changes detected."
+          else
+            echo "changed=true" >> $GITHUB_OUTPUT
+            echo "Report changes detected — will open PR."
+          fi
+
+      - name: Create Pull Request with updated reports
+        if: steps.check-changes.outputs.changed == 'true'
+        uses: peter-evans/create-pull-request@v7
```
Comment (epetrow) [inline .github/workflows/nightly-a11y.yml]: This needs a `permissions` block, consider adding either `contents: write` or `pull-requests: write`

**Comment by zhpenkov** [comment]
This PR has good ideas, but the implementation is heavier than it needs to be for a POC.
Screenshot 2026-03-26 at 12 23 49
Here's what I'd like to see trimmed before considering merging.

### 1. Drop the custom HTML output

The `markdownToHtml()` function is ~70 lines of fragile regex-based conversion. 
See if we need that.

If it is mandatory, use a verified tool.

Use an existing tool (`marked`, `pandoc`, `md-to-pdf`) instead. We don't want to maintain a homebrew Markdown→HTML renderer.

### 2. Drop the `baseline-data.ts` code generation

The build step that embeds JSON into a `.ts` file adds complexity for no benefit — this is a build-time tool, not a browser runtime. Just read the JSON files directly at runtime.

### 3. Don't hardcode the full WCAG criteria catalog

`wcag-criteria.ts` (155 lines) lists every WCAG 2.0/2.1/2.2 criterion. This will drift and become stale. Either:
- Import from `axe-core` (it already ships these mappings), or
- Keep it as a plain JSON file that's easier to diff and update

### 4. Remove the `AAA_DEFAULTS` table

Pre-filling "Not Applicable" / "Supports" for AAA criteria is judgment-heavy and could need legal review. Just use "Not Evaluated" for all AAA — it's the honest answer and zero maintenance.

### 5. Simplify config merging

The exception-merging logic (key by `wcagCriteria|axeRuleId`, replace vs append) is premature optimization. At this stage, there's one consumer — just spread/assign the whole config. Sophisticated merging can come when there's an actual need.

### 6. Remove the nightly workflow

`nightly-a11y.yml` is premature for a POC. It creates noise PRs, needs broad write permissions, and silently swallows failures with `continue-on-error: true`. Add it separately once the generator is proven.

### 7. Simplify the CI path detection

The `check-a11y-paths` job shells out to `gh api`, parses JSON with `jq`, handles truncation at 300 files — all fragile. Replace with the built-in `paths` filter or `dorny/paths-filter`, which handles all of this in one line.

### 8. Remove partial report merging

The ~60 lines added to `test-a11y-unified.mjs` for merging partial runs into an existing report are risky (stale data + partial update = misleading results). For now, write what was tested. Merging can be a separate, deliberate step later.

### 9. Move or simplify the CLI wrapper

`scripts/vpat/generate-vpat.mjs` requires the built output from index.js, meaning the HTML package must be pre-built. Either make it a standalone script that imports the source directly, or put the CLI inside HTML.

### 10. ESLint config change is unrelated

The `unit-tests` config block for `units/**/*.ts` is fine, but doesn't belong in this PR.

### Bottom line

The VPAT generator idea is solid. But the implementation should start as a **simple ~200-300-line script** that reads the config + test JSON and outputs Markdown. Everything else (HTML rendering, nightly CI, partial merging, build-time codegen, AAA pre-filling) can be layered on incrementally once the core is proven.

@telerik/kendo-front-end

**Comment by dtopalov** [comment]
Addressed Review Comments
1. Drop the custom HTML output
Removed the markdownToHtml() function (~70 lines of regex-based conversion). The generator now outputs Markdown only. For HTML/PDF, use npx md-to-pdf. Removed [html](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) field from [VpatReport](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) type.

2. Drop the baseline-data.ts code generation
Deleted baseline-data.ts (65K lines of embedded JSON) and removed the codegen step from [build.js](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html). The generator no longer bundles test data — a11y and contrast reports are passed explicitly via [options.a11yReport](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) / [options.contrastReport](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html).

3. Don't hardcode the full WCAG criteria catalog
Kept [wcag-criteria.ts](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) as it's needed for the VPAT template rows (criteria names, URLs, level annotations), but removed AAA_DEFAULTS (see #4). The [AXE_TO_WCAG](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) mapping is minimal and stable.

4. Remove the AAA_DEFAULTS table
Removed all pre-filled AAA conformance levels. All AAA criteria now default to "Not Evaluated" — the honest answer with zero maintenance.

5. Simplify config merging
Replaced the composite-key ([wcagCriteria|axeRuleId](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)) merging function with simpler inline logic keyed by [wcagCriteria](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) only. Config merging is kept so product repos can add/alter exceptions via overrides.

6. Remove the nightly workflow
Deleted [nightly-a11y.yml](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) (114 lines).

7. Simplify the CI path detection
Replaced the custom gh api + jq shell script (handling truncation, fail-open, etc.) with dorny/paths-filter@v3 — one action that handles everything.

8. Remove partial report merging
Removed ~60 lines of merge logic from [test-a11y-unified.mjs](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html). The script now writes only what was tested in the current run.

9. Simplify the CLI wrapper
Removed [--format](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) flag (Markdown only), simplified the output flow, updated README to match.

10. ESLint config change is unrelated
Reverted the unit-tests ESLint config block and the specificity file lint fixes it triggered. Added --no-warn-ignored to lint-staged in [package.json](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) to suppress the pre-existing "no matching configuration" warnings without introducing a new config block in this PR.

File: tests/_output/a11y-unified-report.json
Comment (zhpenkov) [inline tests/_output/a11y-unified-report.json:1]: That's 56k lines of code JSON, do we need to commit it?

What is the idea? Are we going to ship those?

File: tests/_output/contrast-report.json
Comment (zhpenkov) [inline tests/_output/contrast-report.json:1]: This as well ?

File: packages/html/vpat/default-config.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    specialConsiderations: [
+        'Components that represent or directly extend a native HTML element (e.g., Button, CheckBox, TextBox) do not require additional accessibility functionality to provide their respective level of compliance.',
+        'Components built using other fully accessible components and/or native HTML elements (e.g., Card, AppBar, Pager) achieve compliance through their building blocks.',
+        'Components that do not provide end-user interaction and serve only as visual representation (e.g., Icons, ProgressBars, Badges) are neither focusable nor navigable. Adding descriptive labels or WAI-ARIA attributes to such components is the developer\'s responsibility.',
+        'Assistive technologies treat components like the various Gauges as images. To make them accessible, add a descriptive label or alternative data representation.',
+        'Component templates introduce custom markup that may not be accessible. Test any modifications to ensure accessibility compliance.',
+        'The compliance levels are based on the default component configuration. Enabled features may affect compliance.'
+    ],
+
+    knownExceptions: [
```
Comment (zhpenkov) [inline packages/html/vpat/default-config.ts:33]: Can we catch those exceptions and have them identified programatically, as currently they are just hardcoded?

I usually constrain of hardcoding things as they will eventually drift and the whole comment will be false until someone goes and updates it.

**Comment by dtopalov** [comment]
Closing this PR because:

- component a11y fixes - https://github.com/telerik/kendo-themes/pull/5884 and
- aria specs folder cleanup -https://github.com/telerik/kendo-themes/pull/5888
are moved to separate PRs, 

VPAT generation is extracted in a dedicated repo - https://github.com/telerik/vpat-generator/pull/1
